### PR TITLE
PP-10210 Yourpsp Stripe display task list

### PR DESCRIPTION
**Pull request by AbdirizakIdris**
- Added service tasks for new Stripe onboarding tasklist
- Added unit test for the service tasks
- Updated your Psp controller to include the new tasklist

### CODE REVIEW

File: app/controllers/your-psp/yourpsp-tasks.service.js [hunk 1]
```diff
@@ -0,0 +1,50 @@
+'use strict'
```
Comment (kbottla) [inline app/controllers/your-psp/yourpsp-tasks.service.js:1]: Can this file be renamed to `your-psp/your-psp-tasks.service.js`?

File: app/controllers/your-psp/yourpsp-tasks.service.js [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+      'ENTER_COMPANY_NUMBER': {
+        enabled: !stripeSetupStageComplete(account, 'companyNumber'),
+        complete: stripeSetupStageComplete(account, 'companyNumber')
+      },
+      'CONFIRM_ORGANISATION_DETAILS': {
+        enabled: !stripeSetupStageComplete(account, 'organisationDetails'),
+        complete: stripeSetupStageComplete(account, 'organisationDetails')
+      },
+      'UPLOAD_GOVERNMENT_ENTITY_DOCUMENT': {
+        enabled:  !stripeSetupStageComplete(account, 'governmentEntityDocument'),
```
Comment (kbottla) [inline app/controllers/your-psp/yourpsp-tasks.service.js:38]: Government entity task should only be enabled when all other tasks are completed. Can other tasks be checked?

Could be something like
```
enabled: stripeSetupStageComplete(account, 'bankAccount') &&
          stripeSetupStageComplete(account, 'responsiblePerson') &&
          ...  [rest of the tasks] ...   
          !stripeSetupStageComplete(account, 'governmentEntityDocument')
```

and additional tests

File: app/controllers/your-psp/yourpsp-tasks.service.test.js [hunk 3]
```diff
@@ -0,0 +1,69 @@
+const { expect } = require('chai')
```
Comment (kbottla) [inline app/controllers/your-psp/yourpsp-tasks.service.test.js:1]: Can this be renamed too `app/controllers/your-psp/your-psp-tasks.service.test.js` ?

File: app/controllers/your-psp/yourpsp-tasks.service.js [hunk 4]
```diff
@@ -0,0 +1,50 @@
+'use strict'
+
+function stripeSetupStageComplete (account, stage) {
+  if (account.connectorGatewayAccountStripeProgress) {
+    return account.connectorGatewayAccountStripeProgress[stage]
+  }
+  return false
+}
+
+function getNewStripeTaskList (targetCredential, account) {
```
Comment (kbottla) [inline app/controllers/your-psp/yourpsp-tasks.service.js:10]:
```suggestion
function getStripeTaskList (activeCredential, account) {
```

File: app/controllers/your-psp/yourpsp-tasks.service.js [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+      'UPLOAD_GOVERNMENT_ENTITY_DOCUMENT': {
+        enabled:  !stripeSetupStageComplete(account, 'governmentEntityDocument'),
+        complete: stripeSetupStageComplete(account, 'governmentEntityDocument')
+      }
+    }
+  }
+  throw new Error('Unsupported payment provider')
+}
+
+function stripeTaskListisComplete (taskList) {
```
Comment (kbottla) [inline app/controllers/your-psp/yourpsp-tasks.service.js:46]:
```suggestion
function stripeTaskListIsComplete (taskList) {
```

File: app/controllers/your-psp/yourpsp-tasks.service.test.js [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+const { getCurrentCredential } = require('../../utils/credentials')
+const { getNewStripeTaskList, stripeTaskListisComplete } = require('./yourpsp-tasks.service')
+const gatewayAccountFixtures = require('../../../test/fixtures/gateway-account.fixtures')
+
+describe('Progress of Stripe Psp tasklist page', () => {
```
Comment (kbottla) [inline app/controllers/your-psp/yourpsp-tasks.service.test.js:6]:
```suggestion
describe('Progress of Stripe PSP tasklist', () => {
```

File: app/controllers/your-psp/yourpsp-tasks.service.test.js [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+      responsiblePerson: false,
+      organisationDetails: false,
+      governmentEntityDocument: false
+    }
+    const activeCredential = getCurrentCredential(account)
+    const taskList = getNewStripeTaskList(activeCredential, account)
+    expect(stripeTaskListisComplete(taskList)).to.equal(false)
+  })
+
+  it('correctly calculates progress required for Stripe tasks when one of the conditions is not met', () => {
```
Comment (kbottla) [inline app/controllers/your-psp/yourpsp-tasks.service.test.js:49]:
```suggestion
  it('should correctly calculate progress required for Stripe tasks when one of the conditions is not met', () => {
```

File: app/controllers/your-psp/yourpsp-tasks.service.test.js [hunk 8]
```diff
@@ -0,0 +1,69 @@
+const { expect } = require('chai')
+const { getCurrentCredential } = require('../../utils/credentials')
+const { getNewStripeTaskList, stripeTaskListisComplete } = require('./yourpsp-tasks.service')
+const gatewayAccountFixtures = require('../../../test/fixtures/gateway-account.fixtures')
+
+describe('Progress of Stripe Psp tasklist page', () => {
+  it('correctly calculates all conditions being met for Stripe', () => {
```
Comment (kbottla) [inline app/controllers/your-psp/yourpsp-tasks.service.test.js:7]:
```suggestion
  it('should correctly calculate all conditions being met for Stripe', () => {
```

File: app/controllers/your-psp/yourpsp-tasks.service.test.js [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+      bankAccount: true,
+      director: true,
+      vatNumber: true,
+      companyNumber: true,
+      responsiblePerson: true,
+      organisationDetails: true,
+      governmentEntityDocument: true
+    }
+    const activeCredential = getCurrentCredential(account)
+    const taskList = getNewStripeTaskList(activeCredential, account)
```
Comment (kbottla) [inline app/controllers/your-psp/yourpsp-tasks.service.test.js:24]: Can some assertions be added for the task list which can check if `enabled & complete` values are correctly set for each task?

File: app/controllers/your-psp/yourpsp-tasks.service.test.js [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+      bankAccount: true,
+      director: true,
+      vatNumber: true,
+      companyNumber: true,
+      responsiblePerson: true,
+      organisationDetails: true,
+      governmentEntityDocument: false
+    }
+    const activeCredential = getCurrentCredential(account)
+    const taskList = getNewStripeTaskList(activeCredential, account)
```
Comment (kbottla) [inline app/controllers/your-psp/yourpsp-tasks.service.test.js:66]: Same as above for task list. Can some assertions be added for the task list which can check if `enabled & complete` values are correctly set for each task?

File: app/controllers/your-psp/yourpsp-tasks.service.test.js [hunk 11]
```diff
@@ -0,0 +1,69 @@
...
+      bankAccount: false,
+      director: false,
+      vatNumber: false,
+      companyNumber: false,
+      responsiblePerson: false,
+      organisationDetails: false,
+      governmentEntityDocument: false
+    }
+    const activeCredential = getCurrentCredential(account)
+    const taskList = getNewStripeTaskList(activeCredential, account)
```
Comment (kbottla) [inline app/controllers/your-psp/yourpsp-tasks.service.test.js:45]: Same as above for task list. Can some assertions be added for the task list which can check if `enabled & complete` values are correctly set for each task?

**Review by kbottla — CHANGES_REQUESTED** [review]

File: app/controllers/your-psp/your-psp-tasks.service.test.js [hunk 12]
```diff
@@ -0,0 +1,153 @@
+const { expect } = require('chai')
+const { getCurrentCredential } = require('../../utils/credentials')
+const { getStripeTaskList, stripeTaskListIsComplete } = require('./your-psp-tasks.service')
+const gatewayAccountFixtures = require('../../../test/fixtures/gateway-account.fixtures')
+
+describe('Progress of Stripe PSP tasklist page', () => {
+  it('should correctly calculate all conditions being met for Stripe',  () => {
```
Comment (kbottla) [inline app/controllers/your-psp/your-psp-tasks.service.test.js:7]: Can we be more specific about what we are testing? 

```suggestion
  it('should calculate Stripe tasks statuses correctly when all tasks are completed', () => {
```

File: app/controllers/your-psp/your-psp-tasks.service.test.js [hunk 13]
```diff
@@ -0,0 +1,153 @@
...
+
+    expect(taskList.ENTER_BANK_DETAILS.completed).to.equal(true)
+    expect(taskList.ENTER_RESPONSIBLE_PERSON.completed).to.equal(true)
+    expect(taskList.ENTER_DIRECTOR.completed).to.equal(true)
+    expect(taskList.ENTER_VAT_NUMBER.completed).to.equal(true)
+    expect(taskList.ENTER_COMPANY_NUMBER.completed).to.equal(true)
+    expect(taskList.CONFIRM_ORGANISATION_DETAILS.completed).to.equal(true)
+    expect(taskList.UPLOAD_GOVERNMENT_ENTITY_DOCUMENT.completed).to.equal(true)
+
+    expect(stripeTaskListIsComplete(taskList)).to.equal(true)
```
Comment (kbottla) [inline app/controllers/your-psp/your-psp-tasks.service.test.js:41]:
```suggestion
    expect(stripeTaskListIsComplete(taskList)).to.equal(true)
```

File: app/controllers/your-psp/your-psp-tasks.service.js [hunk 14]
```diff
@@ -0,0 +1,56 @@
...
+
+function stripeSetupStageComplete (account, stage) {
+  if (account.connectorGatewayAccountStripeProgress) {
+    return account.connectorGatewayAccountStripeProgress[stage]
+  }
+  return false
+}
+
+function getStripeTaskList (targetCredential, account) {
+   if (targetCredential.payment_provider === 'stripe') {
```
Comment (kbottla) [inline app/controllers/your-psp/your-psp-tasks.service.js:11]:
```suggestion
  if (targetCredential.payment_provider === 'stripe') {
```

File: app/controllers/your-psp/your-psp-tasks.service.test.js [hunk 15]
```diff
@@ -0,0 +1,153 @@
+const { expect } = require('chai')
+const { getCurrentCredential } = require('../../utils/credentials')
+const { getStripeTaskList, stripeTaskListIsComplete } = require('./your-psp-tasks.service')
+const gatewayAccountFixtures = require('../../../test/fixtures/gateway-account.fixtures')
+
+describe('Progress of Stripe PSP tasklist page', () => {
```
Comment (kbottla) [inline app/controllers/your-psp/your-psp-tasks.service.test.js:6]:
```suggestion
describe('Progress of Stripe PSP tasklist', () => {
```

File: app/controllers/your-psp/your-psp-tasks.service.test.js [hunk 16]
```diff
@@ -0,0 +1,153 @@
...
+    expect(taskList.ENTER_DIRECTOR.completed).to.equal(true)
+    expect(taskList.ENTER_VAT_NUMBER.completed).to.equal(true)
+    expect(taskList.ENTER_COMPANY_NUMBER.completed).to.equal(true)
+    expect(taskList.CONFIRM_ORGANISATION_DETAILS.completed).to.equal(true)
+    expect(taskList.UPLOAD_GOVERNMENT_ENTITY_DOCUMENT.completed).to.equal(true)
+
+    expect(stripeTaskListIsComplete(taskList)).to.equal(true) 
+  })
+
+  it('correctly calculates progress required for Stripe tasks when non of the conditions are met', () => {
```
Comment (kbottla) [inline app/controllers/your-psp/your-psp-tasks.service.test.js:44]: Would be good to start test names with `should` should it reads `it should ...`

```suggestion
  it('should calculate Stripe tasks statuses correctly when none of the tasks are completed', () => {
```

File: app/controllers/your-psp/your-psp-tasks.service.test.js [hunk 17]
```diff
@@ -0,0 +1,153 @@
...
+    expect(taskList.ENTER_DIRECTOR.completed).to.equal(false)
+    expect(taskList.ENTER_VAT_NUMBER.completed).to.equal(false)
+    expect(taskList.ENTER_COMPANY_NUMBER.completed).to.equal(false)
+    expect(taskList.CONFIRM_ORGANISATION_DETAILS.completed).to.equal(false)
+    expect(taskList.UPLOAD_GOVERNMENT_ENTITY_DOCUMENT.completed).to.equal(false)
+    
+    expect(stripeTaskListIsComplete(taskList)).to.equal(false)
+  })
+
+  it('should correctly calculate progress required for Stripe tasks when one of the conditions is not met', () => {
```
Comment (kbottla) [inline app/controllers/your-psp/your-psp-tasks.service.test.js:81]:
```suggestion
  it('should calculate Stripe tasks statuses correctly when one of the tasks are not completed', () => {
```

File: app/controllers/your-psp/your-psp-tasks.service.test.js [hunk 18]
```diff
@@ -0,0 +1,153 @@
...
+    expect(taskList.ENTER_DIRECTOR.completed).to.equal(true)
+    expect(taskList.ENTER_VAT_NUMBER.completed).to.equal(true)
+    expect(taskList.ENTER_COMPANY_NUMBER.completed).to.equal(true)
+    expect(taskList.CONFIRM_ORGANISATION_DETAILS.completed).to.equal(true)
+    expect(taskList.UPLOAD_GOVERNMENT_ENTITY_DOCUMENT.completed).to.equal(false)
+
+    expect(stripeTaskListIsComplete(taskList)).to.equal(false)
+  })
+
+  it('should correctly calculate progress for upload govenment document to be enabled only when all the other conditions are met', () => {
```
Comment (kbottla) [inline app/controllers/your-psp/your-psp-tasks.service.test.js:118]:
```suggestion
  it('should return upload govenment document as enabled only when all the other tasks are completed', () => {
```

File: app/controllers/your-psp/your-psp-tasks.service.test.js [hunk 19]
```diff
@@ -0,0 +1,153 @@
...
+      vatNumber: true,
+      companyNumber: true,
+      responsiblePerson: true,
+      organisationDetails: true,
+      governmentEntityDocument: false
+    }
+    const activeCredential = getCurrentCredential(account)
+    const taskList = getStripeTaskList(activeCredential, account)
+
+    expect(taskList.ENTER_BANK_DETAILS.enabled).to.equal(false)
```
Comment (kbottla) [inline app/controllers/your-psp/your-psp-tasks.service.test.js:136]: As this test is focused on 'upload goverment entity' task, we remove checking for enabled for other tasks.

File: app/controllers/your-psp/your-psp-tasks.service.test.js [hunk 20]
```diff
@@ -0,0 +1,153 @@
...
+
+    expect(taskList.ENTER_BANK_DETAILS.completed).to.equal(true)
+    expect(taskList.ENTER_RESPONSIBLE_PERSON.completed).to.equal(true)
+    expect(taskList.ENTER_DIRECTOR.completed).to.equal(true)
+    expect(taskList.ENTER_VAT_NUMBER.completed).to.equal(true)
+    expect(taskList.ENTER_COMPANY_NUMBER.completed).to.equal(true)
+    expect(taskList.CONFIRM_ORGANISATION_DETAILS.completed).to.equal(true)
+    expect(taskList.UPLOAD_GOVERNMENT_ENTITY_DOCUMENT.completed).to.equal(false)
+
+    expect(stripeTaskListIsComplete(taskList.UPLOAD_GOVERNMENT_ENTITY_DOCUMENT.enabled)).to.equal(true)
```
Comment (kbottla) [inline app/controllers/your-psp/your-psp-tasks.service.test.js:151]: We are not testing method `stripeTaskListIsComplete` here. so assertion should be 

```suggestion
    expect(taskList.UPLOAD_GOVERNMENT_ENTITY_DOCUMENT.enabled).to.equal(true)
```

File: app/controllers/your-psp/your-psp-tasks.service.test.js (hunk 15 above)
Comment (kbottla) [inline app/controllers/your-psp/your-psp-tasks.service.test.js:6]: Every test is checking multiple methods `getStripeTaskList` and `stripeTaskListIsComplete`.

Would be good to have a separate group of tests so it is clear what is being tested. So brought structure could be like below.

```
describe('Stripe task list', () => {

describe('Get Stripe task list status', () => {
   it('should calculate Stripe status (enabled & completed) correctly for each task when completed', () => {
   })
   it('should calculate Stripe status (enabled & completed) correctly for each task when not completed', () => {
   })
   it('should return upload govenment document as enabled only when all the other tasks are completed', () => {
   })
   it('should throw error if payment provider is not Stripe', () => {
   })
})

describe('Calculate Stripe tasks completess', () => {
   it('should return true if all tasks are completed', () => {
   })
   it('should return false if none of the tasks are completed', () => {
   })
   it('should return false if any task is not completed', () => {
   })
})

})
```

Comment (kbottla) [inline app/controllers/your-psp/your-psp-tasks.service.test.js:6]: One more test can be added for the error scenario (i.e., when payment provider is not Stripe)

File: app/controllers/your-psp/yourpsp-tasks.service.js (hunk 1 above)
Comment (AbdirizakIdris) [reply app/controllers/your-psp/yourpsp-tasks.service.js]: Yes done

File: app/controllers/your-psp/yourpsp-tasks.service.js (hunk 2 above)
Comment (AbdirizakIdris) [reply app/controllers/your-psp/yourpsp-tasks.service.js]: Done too

File: app/controllers/your-psp/yourpsp-tasks.service.test.js (hunk 3 above)
Comment (AbdirizakIdris) [reply app/controllers/your-psp/yourpsp-tasks.service.test.js]: Yes, renamed.

File: app/controllers/your-psp/yourpsp-tasks.service.js (hunk 4 above)
Comment (AbdirizakIdris) [reply app/controllers/your-psp/yourpsp-tasks.service.js]: Renamed getNewStripeTaskList to getStripeTaskList .

File: app/controllers/your-psp/yourpsp-tasks.service.test.js (hunk 7 above)
Comment (AbdirizakIdris) [reply app/controllers/your-psp/yourpsp-tasks.service.test.js]: Done

File: app/controllers/your-psp/your-psp-tasks.service.test.js (hunk 13 above)
Comment (AbdirizakIdris) [reply app/controllers/your-psp/your-psp-tasks.service.test.js]: Removed space

File: app/controllers/your-psp/your-psp-tasks.service.test.js (hunk 19 above)
Comment (AbdirizakIdris) [reply app/controllers/your-psp/your-psp-tasks.service.test.js]: Ok, removed checking for other tasks

File: app/controllers/your-psp/your-psp-tasks.service.test.js (hunk 15 above)
Comment (AbdirizakIdris) [reply app/controllers/your-psp/your-psp-tasks.service.test.js]: Yes this seems a better approach.

Comment (AbdirizakIdris) [reply app/controllers/your-psp/your-psp-tasks.service.test.js]: Done

File: app/controllers/your-psp/get.controller.js [hunk 21]
```diff
@@ -29,6 +30,9 @@ module.exports = async (req, res, next) => {
       if (stripeData.requiresAdditionalKycData || kycCompleted) {
         stripeData.kycTaskList = await getTaskList(activeCredential)
         stripeData.kycTaskListComplete = isComplete(stripeData.kycTaskList)
+      } else {
+        stripeData.taskList = yourPspTasks.getStripeTaskList(activeCredential, req.account)
+        stripeData.taskListIsComplete = yourPspTasks.stripeTaskListIsComplete(stripeData.taskList)
```
[the remark this answers is not in the document]
Comment (AbdirizakIdris) [reply app/controllers/your-psp/get.controller.js]: Ok

**Review by kbottla — APPROVED** [review]

**Pull request merged**